File: public/java/031-binary-tree-inorder-traversal.java
/*
 * LeetCode Problem #94: Binary Tree Inorder Traversal
 * Difficulty: Easy
 * Link: https://leetcode.com/problems/binary-tree-inorder-traversal/
 */
import java.util.*;

public class Solution {

    static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }

    // APPROACH 1: Recursive | O(n) time | O(h) space
    // EXPLAIN: Recursively traverse left subtree, visit root, traverse right subtree.
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        dfs(root, result);
        return result;
    }
    private void dfs(TreeNode node, List<Integer> result) {
        if (node == null) return;
        dfs(node.left, result);
        result.add(node.val);
        dfs(node.right, result);
    }

    // APPROACH 2: Iterative with Stack | O(n) time | O(h) space
    // EXPLAIN: Push all left nodes to stack; pop to record and then process right subtree.
    public List<Integer> inorderTraversalIterative(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        Deque<TreeNode> stack = new ArrayDeque<>();
        TreeNode curr = root;
        while (curr != null || !stack.isEmpty()) {
            while (curr != null) { stack.push(curr); curr = curr.left; }
            curr = stack.pop();
            result.add(curr.val);
            curr = curr.right;
        }
        return result;
    }

    // APPROACH 3: Morris Traversal | O(n) time | O(1) space
    // EXPLAIN: Use threaded-tree technique to traverse without recursion or stack; create/remove threads to predecessors.
    public List<Integer> inorderTraversalMorris(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        TreeNode curr = root;
        while (curr != null) {
            if (curr.left == null) {
                result.add(curr.val);
                curr = curr.right;
            } else {
                TreeNode pred = curr.left;
                while (pred.right != null && pred.right != curr) pred = pred.right;
                if (pred.right == null) {
                    pred.right = curr;
                    curr = curr.left;
                } else {
                    pred.right = null;
                    result.add(curr.val);
                    curr = curr.right;
                }
            }
        }
        return result;
    }

    // APPROACH 4: Iterative with Color Marking | O(n) time | O(h) space
    // EXPLAIN: Stack of (node, visited) pairs; on first encounter push right/root-marked/left; on second (marked) encounter record the value.
    public List<Integer> inorderTraversalColor(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) return result;
        Deque<Object[]> stack = new ArrayDeque<>();
        stack.push(new Object[]{root, false});
        while (!stack.isEmpty()) {
            Object[] top = stack.pop();
            TreeNode node = (TreeNode) top[0];
            boolean visited = (boolean) top[1];
            if (node == null) continue;
            if (visited) {
                result.add(node.val);
            } else {
                stack.push(new Object[]{node.right, false});
                stack.push(new Object[]{node, true});
                stack.push(new Object[]{node.left, false});
            }
        }
        return result;
    }
}

// Made with Bob

File: public/python/031-binary-tree-inorder-traversal.py
# LeetCode Problem #94: Binary Tree Inorder Traversal
# Difficulty: Easy
# Link: https://leetcode.com/problems/binary-tree-inorder-traversal/

from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# ─────────────────────────────────────────────
# APPROACH 1: Recursive | O(n) time | O(h) space
# EXPLAIN: Recursively traverse left, visit root, traverse right.
# WHEN: Simplest and most readable; standard first answer.

def inorderTraversal_recursive(root: Optional[TreeNode]) -> List[int]:
    result = []
    def dfs(node):
        if not node:
            return
        dfs(node.left)
        result.append(node.val)
        dfs(node.right)
    dfs(root)
    return result


# ─────────────────────────────────────────────
# APPROACH 2: Iterative with Stack | O(n) time | O(h) space
# EXPLAIN: Use a stack to simulate the call stack: push all left nodes, pop and record, then move right.
# WHEN: Preferred when recursion is not allowed; mirrors the recursive approach explicitly.

def inorderTraversal_iterative(root: Optional[TreeNode]) -> List[int]:
    result = []
    stack = []
    curr = root
    while curr or stack:
        while curr:
            stack.append(curr)
            curr = curr.left
        curr = stack.pop()
        result.append(curr.val)
        curr = curr.right
    return result


# ─────────────────────────────────────────────
# APPROACH 3: Morris Traversal | O(n) time | O(1) space
# EXPLAIN: Create temporary threads from each node's inorder predecessor back to the node; process node when thread is removed.
# WHEN: Optimal space — O(1) extra memory; temporarily modifies tree structure.

def inorderTraversal_morris(root: Optional[TreeNode]) -> List[int]:
    result = []
    curr = root
    while curr:
        if not curr.left:
            result.append(curr.val)
            curr = curr.right
        else:
            # Find inorder predecessor
            pred = curr.left
            while pred.right and pred.right is not curr:
                pred = pred.right
            if not pred.right:
                pred.right = curr      # create thread
                curr = curr.left
            else:
                pred.right = None      # remove thread
                result.append(curr.val)
                curr = curr.right
    return result


# ─────────────────────────────────────────────
# APPROACH 4: Iterative with Color Marking | O(n) time | O(h) space
# EXPLAIN: Use a stack of (node, visited) pairs; push right/node(marked)/left in reverse order; process marked nodes immediately.
# WHEN: Unified template that works for pre/in/post-order by just reordering the push sequence.

def inorderTraversal_color(root: Optional[TreeNode]) -> List[int]:
    result = []
    if not root:
        return result
    stack = [(root, False)]
    while stack:
        node, visited = stack.pop()
        if not node:
            continue
        if visited:
            result.append(node.val)
        else:
            stack.append((node.right, False))
            stack.append((node, True))
            stack.append((node.left, False))
    return result


# Made with Bob
